### PR TITLE
End TreeBuilder's reliance on DOM.

### PR DESCRIPTION
**Pull request by nikhilshagri**
---

Fixed h5e's TreeBuilder so that it does not use `same_tree` and `has_parent_node` methods: https://github.com/cynicaldevil/html5ever/commit/d8c2ea5cb62a1a755bb78b9f5de05d966776883a

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18056)

### CODE REVIEW

**Comment by cynicaldevil** [comment]
cc @nox @jdm

File: components/script/dom/servoparser/async_html.rs [hunk 1]
```diff
@@ -400,7 +398,17 @@ impl Tokenizer {
...
                 }
             }
-            ParseOperation::AssociateWithForm { target, form } => {
+            ParseOperation::AssociateWithForm { target, form, element, prev_element } => {
+                let tree_node = if !prev_element.is_some() ||
+                                   (prev_element.is_some() && self.has_parent_node(element)) {
+                    element
+                } else {
+                    prev_element.expect("Previous element was expected here")
+                };
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]:
```rust
let tree_node = prev_element.map_or(element, |prev| {
    if self.has_parent_node(element) { element } else { prev }
});
```

No need for `unwrap` or `expect`.

File: components/script/dom/servoparser/async_html.rs [hunk 2]
```diff
@@ -633,18 +635,20 @@ impl TreeSink for Sink {
...
-        }
+    fn has_parent_node(&self, _: &Self::Handle) -> bool {
+        unreachable!();
     }
 
-    fn associate_with_form(&mut self, target: &Self::Handle, form: &Self::Handle) {
+    fn associate_with_form(&mut self,
+        target: &Self::Handle,
+        form: &Self::Handle,
+        nodes: (&Self::Handle, Option<&Self::Handle>)) {
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Nit: block indenting, the arguments list closing paren and opening brace should be on their own line.

File: components/script/dom/servoparser/mod.rs [hunk 3]
```diff
@@ -862,6 +874,17 @@ impl TreeSink for Sink {
         insert(&parent, None, child);
     }
 
+    fn append_based_on_parent_node(&mut self,
+        elem: &JS<Node>,
+        prev_elem: &JS<Node>,
+        child: NodeOrText<JS<Node>>) {
```
Comment (nox) [inline components/script/dom/servoparser/mod.rs]: Nit: indentation again.

File: components/script/dom/servoparser/mod.rs [hunk 4]
```diff
@@ -820,7 +820,19 @@ impl TreeSink for Sink {
...
 
-    fn associate_with_form(&mut self, target: &JS<Node>, form: &JS<Node>) {
+    fn associate_with_form(&mut self, target: &JS<Node>, form: &JS<Node>, nodes: (&JS<Node>, Option<&JS<Node>>)) {
+        let (element, prev_element) = nodes;
+        let tree_node = if !prev_element.is_some() ||
+                           (prev_element.is_some() && self.has_parent_node(element)) {
+            element
+        } else {
+            prev_element.expect("Previous element was expected here")
+        };
```
Comment (nox) [inline components/script/dom/servoparser/mod.rs]: Same remark about `unwrap` and `expect`.

File: components/script/dom/servoparser/async_html.rs [hunk 5]
```diff
@@ -658,6 +662,21 @@ impl TreeSink for Sink {
         self.send_op(ParseOperation::AppendBeforeSibling { sibling: sibling.id, node: new_node });
     }
 
+    fn append_based_on_parent_node(&mut self,
+        elem: &Self::Handle,
+        prev_elem: &Self::Handle,
+        child: HtmlNodeOrText<Self::Handle>) {
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Nit: indentation here.

**Review by nox — CHANGES_REQUESTED** [review]
This looks great!

File: components/script/dom/servoparser/async_html.rs [hunk 6]
```diff
@@ -583,12 +586,8 @@ impl TreeSink for Sink {
         target.qual_name.as_ref().expect("Expected qual name of node!").expanded()
     }
 
-    fn same_tree(&self, x: &Self::Handle, y: &Self::Handle) -> bool {
-        self.sender.send(ToTokenizerMsg::IsSameTree(x.id, y.id)).unwrap();
-        match self.receiver.recv().expect("Unexpected channel panic in html parser thread.") {
-            ToSinkMsg::IsSameTree(result) => result,
-            _ => unreachable!(),
-        }
+    fn same_tree(&self, _: &Self::Handle, _: &Self::Handle) -> bool {
```
Comment (nox) [inline components/script/dom/servoparser/async_html.rs]: Why does this still exist?

**Comment by jdm** [comment]
@bors-servo: retry
* git

**Pull request merged**